### PR TITLE
Make vectors RefCounted

### PR DESCRIPTION
**Pull request by alex-spies**
This allows to share vectors among multiple blocks and will be useful for improving the performance of expand and filter operators.

Relates https://github.com/elastic/elasticsearch/issues/100528.

### CODE REVIEW

File: x-pack/plugin/esql/compute/src/main/generated-src/org/elasticsearch/compute/data/ConstantBooleanVector.java [hunk 1]
```diff
@@ -70,13 +70,4 @@ public int hashCode() {
...
         return getClass().getSimpleName() + "[positions=" + getPositionCount() + ", value=" + value + ']';
     }
-
-    @Override
-    public void close() {
-        if (released) {
-            throw new IllegalStateException("can't release already released vector [" + this + "]");
-        }
-        released = true;
-        blockFactory().adjustBreaker(-ramBytesUsed(), true);
```
Comment (alex-spies) [inline x-pack/plugin/esql/compute/src/main/generated-src/org/elasticsearch/compute/data/ConstantBooleanVector.java:80]: The super method already does exactly this.

File: x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/AbstractThreadLocalRefCounted.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+package org.elasticsearch.compute.data;
+
+import org.elasticsearch.core.RefCounted;
+import org.elasticsearch.core.Releasable;
+
+/**
+ * Releasable, non-threadsafe version of {@link org.elasticsearch.core.AbstractRefCounted}.
+ * Calls to {@link AbstractThreadLocalRefCounted#decRef()} and {@link AbstractThreadLocalRefCounted#close()} are equivalent.
+ */
+abstract class AbstractThreadLocalRefCounted implements RefCounted, Releasable {
```
Comment (nik9000) [inline x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/AbstractThreadLocalRefCounted.java:17]: Might call it `AbstractNonThreadSafeRefCounted` - `ThreadLocal` implies to me that it's a java `ThreadLocal` which has all kinds of fun baggage.

**Review by nik9000 — APPROVED** [review]

**Comment by alex-spies** [comment]
Thanks for the quick review @nik9000 !

**Comment by dnhatn** [comment]
Thank you, Alex!

**Pull request merged**